firedrake/assemble.py

```diff
@@ -2103,21 +2085,45 @@ def index_function_spaces(form, indices):
         else:
             raise AssertionError
 
-    @staticmethod
-    def index_tensor(tensor, form, indices, diagonal):
-        """Return the PyOP2 data structure tied to ``tensor``, indexed
-        if necessary.
-        """
-        rank = len(form.arguments())
-        is_indexed = any(i is not None for i in indices)
 
-        if rank == 0:
-            return tensor
-        elif rank == 1 or rank == 2 and diagonal:
+def _as_pyop2_tensor(tensor, indices=None):
+    """Cast a Firedrake tensor into a PyOP2 data structure, optionally indexing it."""
+    is_indexed = indices and any(index is not None for index in indices)
+
+    if isinstance(tensor, op2.Global):
+        assert not is_indexed
+        return tensor
+
+    if isinstance(tensor, (firedrake.Function, firedrake.Cofunction)):
+        if is_indexed:
             i, = indices
-            return tensor.dat[i] if is_indexed else tensor.dat
-        elif rank == 2:
+            return tensor.dat[i]
+        else:
+            return tensor.dat
+    else:
+        assert isinstance(tensor, firedrake.Matrix)
+        if is_indexed:
             i, j = indices
-            return tensor.M[i, j] if is_indexed else tensor.M
+            return tensor.M[i, j]
         else:
-            raise AssertionError
+            return tensor.M
+
+
+def _as_parloop_type(tensor, indices):
+    """Cast a Firedrake tensor into a PyOP2 data structure suitable for a parloop.
+
+    This function differs from `_as_pyop2_tensor` in that matrices with ``"python"``
+    type are unpacked into their underlying `op2.Global` or `op2.Dat`.
+
+    """
+    pyop2_tensor = _as_pyop2_tensor(tensor, indices)
+
+    if isinstance(pyop2_tensor, op2.Mat) and pyop2_tensor.handle.getType() == "python":
+        mat_context = pyop2_tensor.handle.getPythonContext()
+        if isinstance(mat_context, _GlobalMatPayload):
+            pyop2_tensor = mat_context.global_
+        else:
+            assert isinstance(mat_context, _DatMatPayload)
+            pyop2_tensor = mat_context.dat
+
+    return pyop2_tensor
```

Comment:
This free function should also be implemented as a method of `ParloopFormAssembler`, I think.